r1cs/gadgets/crypto/src/nizk/groth16/mod.rs

```diff
@@ -149,7 +149,9 @@ for Groth16VerifierGadget<PairingE, ConstraintF, P>
                 .enumerate()
                 {
                     let input_bits = input.to_bits(cs.ns(|| format!("Input {}", i)))?;
-                    g_ic = b.mul_bits(cs.ns(|| format!("Mul {}", i)), &g_ic, input_bits.iter())?;
+                    g_ic = b
+                        .mul_bits(cs.ns(|| format!("Mul {}", i)), input_bits.iter())?
```

Comment:
Variable base Groth16 verifier circuit might be a case in which a complete mul_bits is needed. 
However, I wouldn't push that now, as most likely we will purge Groth from ginger.

Reply:
Shall we put a TODO ? And why this use case in particular ?

Reply:
As discussed offline, it is about whether or not scalars are output of a pseudo-random source (a hash, or squeezing from a Fiat-Shamir sponge). 
- If they are from a pseudo-random source, then the probability that they cause an exceptional case not covered by the incomplete circuit is negligible. Consequently an attacker would need to sample such randomness in (expected) superpolynomial time to find it.
- If they are not from a pseudo-random source, then the attacker might craft scalars that meet an exceptional case.

For SNARK (Groth, or our recursive Darlin) the verifier does scalar multiplication on the circuit inputs, which are typically not from a pseudo-random source. 
We can put a TODO addressing completeness (w.r.t. the scalar) in `mul_bits()` and `mul_bits_fixed_base()`.

Reply:
Well let's keep it in mind for ourselves, as this issue might be met only in our protocol where external entities are responsible to create proofs upon receiving data from outside. I'll mark this as resolved.